### ISSUE TITLE
host_machine cpu can change from 64-bit to 32-bit mid-way through setup

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#7533]
There are a bunch of different issues here, and I am not sure which ones are bugs and which ones are features. I wasn't able to come up with a straightforward fix, so I'm going to just describe how this happens.

## CPU family detection has bad assumptions

`environment.py:detect_cpu_family()` detects the CPU family based on information gathered from the system, and from the list of compilers passed in. The first part is fine. The second part is a bit iffy, because if there's a single compiler in that list that outputs 32-bit, we will detect the arch as 32-bit.

This is usually not a problem because no one is trying to use a 32-bit C compiler and a 64-bit C++ compiler, but we could warn or error out when this happens. This is harder than it looks because we do not actually detect the arch of a compiler when we detect it because we don't need to. Exception: we do it on Windows to pass the correct `/MACHINE:` argument to the linker.

## `host_machine` values are created afresh when `project()` or `add_languages()` is called

Each subproject re-detects compilers and generates new values of `host_machine` by calling `detect_cpu_family()`. Thanks to the CPU detection I described above, this can cause a subproject to have a different value of `host_machine.cpu_family()`.

This is definitely wrong. Subprojects must always have the same `host_machine` values as the parent project.

## The list of compilers is in coredata and shared between subprojects

If a subproject does `add_languages()`, all future subprojects will use this shared list of compilers for detection of, you guessed it, the CPU family.

---

This manifests like so: https://gitlab.freedesktop.org/gstreamer/gst-build/-/issues/88

On Windows people have all kinds of garbage in `PATH`. So they can quite easily have C/C++ compilers that are 64-bit MSVC and then suddenly also ObjC/ObjC++ compilers that are 32-bit MinGW-GCC, and if a subproject does `add_languages()`, all later subprojects will think we're building for 32-bit.

This is less likely on other platforms, but can still happen.

CC: @dcbaker and @Ericson2314

**Comment by Ericson2314** [comment]
Yes I think our current approach of interleaved guessing the platforms and finding compilers is quite chaotic. I made all the `default_missing` and `miss_defaulting` stuff that does this, but I never liked it and only made it because the existing algorithm already worked that way (just a bit obfuscated).
